stackslib/src/net/chat.rs

```diff
@@ -2687,6 +2696,118 @@ impl ConversationP2P {
         }
     }
 
+    /// Are we trying to resolve DNS?
+    pub fn waiting_for_dns(&self) -> bool {
+        self.dns_deadline < u128::MAX
+    }
+
+    /// Attempt to resolve the hostname of a conversation's data URL to its IP address.
+    fn try_resolve_data_url_host(
+        &mut self,
+        dns_client_opt: &mut Option<&mut DNSClient>,
+        dns_timeout: u128,
+    ) {
```

Comment:
In this specific case, no, that would not be ideal.

We need the peer's data URL host to be resolved at some point before block-download can begin, but there aren't many good options for this with Rust's native DNS resolution code.  Naively calling `.to_socket_addrs()` (which is synchronous) could cause the entire p2p thread to stall while waiting for a DNS round-trip, which can take an unbounded amount of time.  A faulty DNS server or even unsuitable default DNS resolution options locally could lead to a denial of service.

What we'd ideally have is a DNS resolution state machine, which the p2p state machine could just drive.  We'd furnish it with callbacks to instantiate and tear down UDP (but also TCP!) sockets so we could poll them, and it would handle the nuts and bolts of packet rx/tx for these sockets and would spit out resolved DNS names as they arrive.  However, building this out ourselves would be very time-consuming and would take substantial effort to make fully RFC-compliant.

Our networking stack does the next-best-thing --  it uses a separate DNS thread which carries out `.to_socket_addrs()` calls on its clients' behalf (so, we get to use a ready-made RFC-compliant DNS client), and spits out resolved DNS names via a poll-able handle.  Now, the p2p state machine can simply poll this thread for name resolutions without blocking everything on a DNS round-trip.

The downside to this approach is that there's no easy way to honor the host's DNS cache policy, since this is enforced internally by `getaddrinfo(3)`.  This is a long-standing limitation with `getaddrinfo(3)`, and is what prompted the likes of [c-ares](https://c-ares.org/) to be written.  The caller doesn't know in advance if they're going to get an immediate response from a cached name resolution, or if they're going to be blocked on waiting for a DNS round-trip (which as mentioned is problematic for us).

If you're adamant about using `.to_socket_addrs()` for each and every name resolution, then I suppose I could refactor this PR so that block-download happens in a separate thread altogether, in which case it could safely perform a blocking `.to_socket_addrs()` without stalling the whole networking thread.  Another option is to do what `tokio` does, which is to spawn a thread for each DNS resolution that would require a DNS query, and have the thread call `.to_socket_addrs()` (the `.await` on the `tokio` DNS resolution future literally wraps a `JoinHandle` for the thread).

Reply:
ping @kantai -- is it okay to leave the DNS code as-s, in light of this explanation?

Reply:
Yeah, that's fine